client/quantum_serverless/core/program.py

```diff
@@ -71,22 +92,130 @@ def save_program(self, program: Program) -> bool:
 
-    def get_program(self, name: str, **kwargs) -> Program:
+    def get_program(self, program: str, **kwargs) -> Optional[Program]:
         """Returns program by name of other query criterieas.
 
         Args:
-            name: name of program
+            program: name of the program
             **kwargs: other args
 
         Returns:
             Program
         """
         raise NotImplementedError
+
+
+class ProgramRepository(ProgramStorage):
+    """ProgramRepository."""
+
+    def __init__(
+        self,
+        host: Optional[str] = None,
+        port: Optional[int] = None,
+        token: Optional[str] = None,
+        root: Optional[str] = None,
+    ):
+        """Program repository implementation.
+
+        Args:
+            host: host of backend
+            port: port of backend
+            token: authentication token
+            root: path to directory where program files will be stored
+        """
+        self.root = root or os.path.dirname(os.path.abspath(__file__))
+        self._host = host or os.environ.get(REPO_HOST_KEY, "http://localhost")
+        self._port = port or os.environ.get(REPO_PORT_KEY, 80)
+        self._token = token
+        self._base_url = f"{self._host}:{self._port}/v1/api/nested-programs/"
+
+    def save_program(self, program: Program) -> bool:
+        raise NotImplementedError("Not implemented yet.")
+
+    def get_programs(self, **kwargs) -> List[str]:
+        result = []
+        response = requests.get(url=self._base_url, params=kwargs, timeout=10)
+        if response.ok:
+            response_data = json.loads(response.text)
+            result = [entry.get("title") for entry in response_data.get("results", [])]
+        return result
+
+    def get_program(self, program: str, **kwargs) -> Optional[Program]:
+        result = None
+        response = requests.get(
+            url=f"{self._base_url}",
+            params={"name": program},
+            allow_redirects=True,
+            timeout=10,
+        )
+        if response.ok:
+            response_data = json.loads(response.text)
+            results = response_data.get("results", [])
+            if len(results) > 0:
+                artifact = results[0].get("artifact")
+                result = Program.from_json(results[0])
+                result.working_dir = download_and_unpack_artifact(
+                    artifact_url=artifact, program=result, root=self.root
+                )
+            else:
+                logging.warning("No entries were found for your request.")
+        return result
+
+
+def download_and_unpack_artifact(
+    artifact_url: str,
+    program: Program,
+    root: str,
+    headers: Optional[Dict[str, Any]] = None,
+) -> str:
+    """Downloads and extract artifact files into destination folder.
+
+    Args:
+        artifact_url: url to get artifact from
+        program: program object artifact belongs to
+        root: root of programs folder a.k.a unpack destination
```

Comment:
Similar to my previous feedback about `root`. Totally my opinion but I would change it here by `path` or something more specific/clearer.